docs/decisions/0022-planner-telemetry-enhancement.md

```diff
@@ -0,0 +1,225 @@
+---
+status: { proposed }
+contact: { TaoChenOSU }
+date: { 2023-11-18 }
+deciders: alliscode, dmytrostruk, markwallace, SergeyMenshykh, stephentoub
+consulted: {}
+informed: {}
+---
+
+# Planner Telemetry Enhancement
+
+## Context and Problem Statement
+
+It would be extremely beneficial for applications using Semantic Kernel' planning features to be able to continuously monitor the performance of planners and plans.
+
+## Scenarios
+
+Contoso is a company that is developing an AI application using SK.
+
+1. Contoso needs to continuously monitor the token usage of a particular planner, including prompt tokens, completion tokens, and the total tokens.
+2. Contoso needs to continuously monitor the time it takes for a particular planner to create a plan.
+3. Contoso needs to continuously monitor the success rate of a particular planner in creating a valid plan.
+4. Contoso needs to continuously monitor the success rate of a particular plan type being executed successfully.
+5. Contoso wants to be able to see the token usage of a particular planner run.
+6. Contoso wants to be able to see the time taken to create a plan of a particular planner run.
+7. Contoso wants to be able to see the steps in a plan.
+8. Contoso wants to be able to see the inputs&outputs of each plan step.
+9. Contoso wants to change a few settings that may affect the performance of the planners. They would like to know how the performance will be affected before committing the changes.
+10. Contoso wants to update to a new model that is cheaper and faster. They would like to know how the new model performs in planning tasks.
+
+## Out of scope
+
+1. We focus on Application Insights. Although other telemetry service options are supported technically, we will not cover possible ways of setting them up in this ADR.
```

Comment:
> We focus on Application Insights. [](http://example.com/codeflow?start=2&length=34)

I am wondering if you talked to Azure observability team about that and get their feedback? I am cc'ing a couple of contacts who will be good to sync up about that. 

@reyang @cijothomas @CodeBlanch @utpilla

Reply:
My suggestion is to leverage .NET's built-in APIs for logs,metrics,traces, which is also the OpenTelemetry's recommended API. These APIs are vendor-neutral, and final destination of telemetry can be chosen and changed without any re-instrumentation. [ApplicationInsights officially support this](https://learn.microsoft.com/en-us/azure/azure-monitor/app/opentelemetry-enable?tabs=aspnetcore), but so is almost all other Observability vendors!

Reply:
agree with @cijothomas here. Using the native .NET APIs will allow using ApplicationInsights but not be restricted to it.

Reply:
Yes, we are planning to leverage the System.Diagnostics APIs. The item here just states that we will only show (in an example) how to configure SK to send telemetry to Application Insights but not all destinations.